### PR TITLE
args.threshold-pred-score \n replacement_using BERT added

### PR DESCRIPTION
**Pull request by Suho0922**
In arg --threshold-pred-score is added 
after important score process is updated with 
replacement_using BERT method

### CODE REVIEW

File: utils/attack.py [hunk 1]
```diff
@@ -76,7 +198,7 @@ def run_attack(args, processor, example, feature, pretrained_model, finetuned_mo
     word_predictions = word_predictions[1:-1, :]  # except  [CLS], [SEP]
     # Top-K 개를 뽑아서 가장 높은 스코어 순으로 정렬하며, 가장 plausible 한 예측값들의 모음
     # torch.return_types.topk(values=tensor([5., 4., 3.]), indices=tensor([4, 3, 2]))
-    word_pred_score, word_pred_idx = torch.topk(
+    word_pred_scores_all, word_predictions = torch.topk(
```
Comment (Sunkyoung) [inline utils/attack.py]: torch.topk 에서 return 하는 것이 value랑 indices 여서 변수명을 가독성 좋게 변경했었습니다.
legacy 변수명 그대로 안따라가도 될 것 같아요!

File: utils/attack.py [hunk 2]
```diff
@@ -51,6 +51,128 @@ def get_important_scores(
...
+        #    continue
+        ############################
+        ##  Maybe useless ##########
+        ############################
+        if top_index > args.max_seq_length - 2:
+            continue
+        ############################
+        
+        substitutes = word_predictions[top_index].unsqueeze(0)  # L, k
+        word_pred_scores = word_pred_scores_all[top_index].unsqueeze(0)
```
Comment (Sunkyoung) [inline utils/attack.py]: important score는 CLS와 SEP 제외하고 진행해서 word_index_with_I_score 가 두 special token을 제외한 index들의 list인데 tgt_word 에 top_index + 1는 어떤 이유에서 해주는 걸까요?

Comment (Sunkyoung) [reply utils/attack.py]: 그리고 아래의 word_predictions, word_pred_scores_all은 다르게 top_index으로 get 했는지 이유를 알 수 있을까요?

File: utils/attack.py [hunk 3]
```diff
@@ -51,6 +51,128 @@ def get_important_scores(
...
+
+            #Replace word and check whether the attack is successful
+            temp_replace = final_words
+            temp_replace[top_index+1] = substitute # replace token
+
+            input_ids = torch.tensor(temp_replace)
+            temp_prob = tgt_model(input_ids)[0].squeeze()
+            feature.query += 1
+            temp_prob = torch.softmax(temp_prob, -1)
+            temp_label = torch.argmax(temp_prob)
```
Comment (Sunkyoung) [inline utils/attack.py]: label prediction 하는 다른 코드와 형식을 통일하면 좋을 것 같아요!

Comment (Sunkyoung) [reply utils/attack.py]: 이처럼 수정하는 편이 어떨까요?

```
input_tensor = processor.get_tensor(temp_replace).to('cuda')    

with torch.no_grad():       
    logit = finetuned_model(input_tensor)[0] 
temp_logit= logit.detach()  
temp_label = torch.argmax(pred_logit, dim=1).flatten()  

feature.query += 1
```

**Review by Sunkyoung — CHANGES_REQUESTED** [review]
한 번 커밋하시는 것보다.. 변경 사항마다 (어렵다면 함수 단위라도..) 커밋 해주시면 보기에 훨씬 편리합니당.. :)

File: utils/attack.py (hunk 1 above)
Comment (Suho0922) [reply utils/attack.py]: word_pred_scores_all 은 이대로 유지하는게 좋을 것 같습니다! 
밑에 substitute를 가져오는 과정에서 
word_pred_scores라는 변수에 tgt_word에 해당하는 pred_score를 저장해주기 때문에 이렇게 하는 편이 헷갈리지 않을 것 같습니다 
word_pred_idx 는 혹시 idx 가 무엇을 의미하는지 알 수 있을까요? indices 인가요? 
그렇다면 이건 저렇게 수정해보도록 하겠습니다!

File: utils/attack.py (hunk 2 above)
Comment (Suho0922) [reply utils/attack.py]: 아 넵 이부분에 대해서 설명해드리면 
feature 에 저장되어 있는 inputs_ids 에는 CLS 가 포함 되어 있고 
important score 에는 special token 이 없기 때문에 CLS 가 포함 되어 있지 않습니다 
따라서 important score의 index 는 0 이 첫 단어 가 되지만 
inputs_ids 에서는 CLS 가  첫 단어가 되므로 index top_index 에 1을 더해주어야 합니다 
또 아래의 word_predictions 와 word_pred_scores_all 은 special token 이 없으므로 top index 를 그대로 사용해주어도 됩니다!

File: utils/attack.py (hunk 3 above)
Comment (Suho0922) [reply utils/attack.py]: 오호 이렇게 할 수 있었군요 수정해서 commit 하겠습니다! temp prob 도 필요하기 때문에 그부분 추가하도록 하겠습니다!

**Comment by Suho0922** [comment]
label prediction form is unified 
word_predictions variable changed to word_pred_idx

File: utils/attack.py (hunk 1 above)
Comment (Sunkyoung) [reply utils/attack.py]: 넵 index를 Idx로 줄여서 보통 많이 사용합니다!

File: utils/attack.py (hunk 2 above)
Comment (Sunkyoung) [reply utils/attack.py]: 그러면 이렇게 수정해서 통일하는 건 어떨까요? 
```
target_input = feature.input_ids[1:-1]
tgt_word = target_input [top_index]
```

File: utils/attack.py [hunk 4]
```diff
@@ -109,11 +109,16 @@ def replacement_using_BERT(feature, current_prob, output,pred_label, word_index_
...
 
-            input_ids = torch.tensor(temp_replace)
-            temp_prob = tgt_model(input_ids)[0].squeeze()
+
+            input_tensor = processor.get_tensor(temp_replace).to('cuda')    
+
+            with torch.no_grad():       
+                logit = finetuned_model(input_tensor)[0] 
+            temp_logit= logit.detach() 
+            temp_prob = torch.softmax(temp_logit, -1)
```
Comment (Sunkyoung) [inline utils/attack.py:118]: temp logit이 원래의 temp_prob와 같은 값이라 temp_prob을 추가 정의안해도 되지 않을까 싶습니다!

File: utils/attack.py (hunk 2 above)
Comment (Suho0922) [reply utils/attack.py]: 그것도 좋은 방법입니다 
하지만 이후에 inference 를 진행할때는 앞 뒤에 CLS 와 SEP을 따로 추가 해주어야 합니다
통일하고 CLS 와 SEP을 추가하는 방향으로 수정해 볼까요?

File: utils/attack.py (hunk 4 above)
Comment (Suho0922) [reply utils/attack.py:118]: 아 logit 값이 softmax 되어진 값인가요? 
softmax 되어진 값이 이후에 쓰이기 때문에 따로 적었습니다!

Comment (Sunkyoung) [reply utils/attack.py:118]: 음 softmax 되기 전의 값은 맞는데, predicted label을 위해 사용되는 것이라면, 
`temp_label = torch.argmax(pred_logit, dim=1).flatten() `
이 부분에서 argmax 시에 dimension 1로 한 후 flatten 하기 때문에 따로 softmax를 사용하지 않아도 됩니다!

Comment (Suho0922) [reply utils/attack.py:118]: 앗 넵 이후에 밑에서 softmax 된 값이 gap 계산에 쓰이기 때문에 추가했습니다!

**Review by Sunkyoung — APPROVED** [review]
